rel/server/overlays/bin/start_runtime.exs

```diff
@@ -6,10 +6,22 @@ File.cd!(System.user_home!())
   dist_port: dist_port
 } = System.fetch_env!("LIVEBOOK_RUNTIME") |> Base.decode64!() |> :erlang.binary_to_term()
 
-# This is the only Fly-specific part of starting Livebook as runtime
-app = System.fetch_env!("FLY_APP_NAME")
-machine_id = System.fetch_env!("FLY_MACHINE_ID")
-node = :"#{node_base}@#{machine_id}.vm.#{app}.internal"
+node =
+  cond do
+    System.get_env("FLY_APP_NAME") ->
+      # This is the only Fly-specific part of starting Livebook as runtime
+      app = System.fetch_env!("FLY_APP_NAME")
+      machine_id = System.fetch_env!("FLY_MACHINE_ID")
+      :"#{node_base}@#{machine_id}.vm.#{app}.internal"
+
+    System.get_env("POD_IP") ->
+      # This is the only K8s-specific part of starting Livebook as runtime
+      hostname = System.fetch_env!("POD_IP")
+      :"#{node_base}@#{hostname}"
+
+    true ->
+      raise "expected either POD_IP (for k8s) or FLY_APP_NAME (for Fly.io) to be set"
+  end
```

Comment:
We can make it more generic:

```elixir
node =
  node_template
  |> String.replace(~r/\{\w+?\}/, fn substitute ->
    env = binary_slice(substitute, 1..-2//1)
    System.fetch_env!(env)
  end)
  |> String.to_atom()
```

The above replaces any `{ENV_NAME}` with the env var value.

Then, instead of passing `node_base`, we pass `node_template` as `"#{node_base}@{FLY_MACHINE_ID}.vm.{FLY_APP_NAME}.internal"` and `"#{node_base}@{POD_IP}"` respectively.

@josevalim wdyt?

Reply:
Last time I was arguing to make it more generic and you wanted to keep it simple, and I think this time it will be the opposite. I believe the env vars are fine as and with less indirection, I would be glad to ship it as is. :)

Reply:
We discussed offline, @mruoss let's keep it as is :)